File: test_regress/t/t_dpi_open_dynamic_oob_bad.cpp
// DESCRIPTION: Verilator: Invalid element indices into dynamic DPI open arrays
//
// This file ONLY is placed under the Creative Commons Public Domain.
// SPDX-FileCopyrightText: 2026 Wilson Snyder
// SPDX-License-Identifier: CC0-1.0

#include "Vt_dpi_open_dynamic_oob_bad__Dpi.h"
#include "svdpi.h"

#include <cassert>

void dpi_dynamic_oob(const svOpenArrayHandle values, const svOpenArrayHandle empty_values) {
    assert(svGetArrElemPtr2(values, 1, 0) == nullptr);
    assert(svGetArrElemPtr2(values, 2, 1) == nullptr);
    assert(svGetArrElemPtr2(values, 3, -1) == nullptr);
    assert(svGetArrElemPtr2(values, 4, 3) == nullptr);
    assert(svGetArrElemPtr1(empty_values, 0) == nullptr);
    assert(svGetArrElemPtr1(empty_values, -1) == nullptr);
}

File: test_regress/t/t_dpi_open_dynamic_oob_bad.py
#!/usr/bin/env python3
# DESCRIPTION: Verilator: Invalid element indices into dynamic DPI open arrays
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of either the GNU Lesser General Public License Version 3
# or the Perl Artistic License Version 2.0.
# SPDX-FileCopyrightText: 2026 Wilson Snyder
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

import vltest_bootstrap

test.scenarios('vlt', 'vltmt')
test.compile(threads=2 if test.vltmt else 1, v_flags2=['t/t_dpi_open_dynamic_oob_bad.cpp'])
test.execute(expect_filename=test.golden_filename)
test.passes()
